File: Project1-Clock-OC/Project1-Clock-OC/ViewController.m
//
//  ViewController.m
//  Day1 Clock
//
//  Created by YangJing on 2018/3/5.
//  Copyright © 2018年 YangJing. All rights reserved.
//

#import "ViewController.h"

@interface ViewController () <UITableViewDelegate, UITableViewDataSource>

@property (nonatomic, strong) UILabel *mainTimerLabel;

@property (nonatomic, strong) UIButton *startBtn;

@property (nonatomic, strong) UIButton *resetBtn;

@property (nonatomic, strong) dispatch_source_t timer;

@property (nonatomic, assign) long timerCount;

@property (nonatomic, strong) UITableView *tableView;

@property (nonatomic, strong) NSMutableArray *dataArray;

@property (nonatomic, assign) long currentCount;

@property (nonatomic, assign) long maxCount;

@property (nonatomic, assign) long minCount;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
    [self configNavigationBar];
    [self configSubview];
    
    //    [self equalWidthFont];
}

//找等宽字体，要不显示器总是闪动（1最窄，4最宽）
- (void)equalWidthFont {
    NSArray *fontArray = [UIFont familyNames];
    for (NSString *fontName in fontArray) {
        CGFloat width1 = [@"11" boundingRectWithSize:CGSizeMake(MAXFLOAT, [UIFont fontWithName:fontName size:68].lineHeight) options:NSStringDrawingUsesFontLeading | NSStringDrawingUsesLineFragmentOrigin attributes:@{NSFontAttributeName:[UIFont fontWithName:fontName size:68]} context:nil].size.width;
        CGFloat width2 = [@"44" boundingRectWithSize:CGSizeMake(MAXFLOAT, [UIFont fontWithName:fontName size:68].lineHeight) options:NSStringDrawingUsesFontLeading | NSStringDrawingUsesLineFragmentOrigin attributes:@{NSFontAttributeName:[UIFont fontWithName:fontName size:68]} context:nil].size.width;
        if (width1 == width2) {
            NSLog(@"yangjing_timer: %@", fontName);
        }
    }
    
}

//MARK: - timer
- (void)timerStart {
    [self.startBtn setTitle:@"停止" forState:UIControlStateNormal];
    self.startBtn.backgroundColor = [UIColor colorWithRed:1 green:0 blue:0 alpha:0.3];
    [self.startBtn setTitleColor:[UIColor redColor] forState:UIControlStateNormal];
    [self.startBtn removeTarget:self action:@selector(timerStart) forControlEvents:UIControlEventTouchUpInside];//???
    
    [self.startBtn addTarget:self action:@selector(timerStop) forControlEvents:UIControlEventTouchUpInside];
    
    self.resetBtn.enabled = YES;
    [self.resetBtn setTitle:@"计次" forState:UIControlStateNormal];
    self.resetBtn.backgroundColor = [UIColor colorWithWhite:1 alpha:0.6];
    [self.resetBtn setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    [self.resetBtn removeTarget:self action:@selector(timerReset) forControlEvents:UIControlEventTouchUpInside];//???
    
    [self.resetBtn addTarget:self action:@selector(timerRecord) forControlEvents:UIControlEventTouchUpInside];
    
    dispatch_queue_t queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0);
    self.timer = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, queue);
    dispatch_source_set_timer(self.timer, dispatch_walltime(NULL, 0), 0.01*NSEC_PER_SEC, 0);
    dispatch_source_set_event_handler(self.timer, ^{
        self.timerCount ++;
        self.currentCount ++;
        dispatch_async(dispatch_get_main_queue(), ^{
            self.mainTimerLabel.text = [NSString stringWithFormat:@"%.2ld:%.2ld.%.2ld", self.timerCount/100/60, self.timerCount/100%60, self.timerCount%100];
        });
    });
    dispatch_resume(self.timer);
}

- (void)timerStop {
    if (self.timer) {
        dispatch_source_cancel(self.timer);
        self.timer = nil;
    }
    
    [self.startBtn setTitle:@"开始" forState:UIControlStateNormal];
    self.startBtn.backgroundColor = [UIColor colorWithRed:0 green:1 blue:0 alpha:0.3];
    [self.startBtn setTitleColor:[UIColor greenColor] forState:UIControlStateNormal];
    [self.startBtn removeTarget:self action:@selector(timerStop) forControlEvents:UIControlEventTouchUpInside];
    
    [self.startBtn addTarget:self action:@selector(timerStart) forControlEvents:UIControlEventTouchUpInside];
    
    if (self.timerCount > 0) {
        [self.resetBtn setTitle:@"复位" forState:UIControlStateNormal];
        self.resetBtn.backgroundColor = [UIColor colorWithWhite:1 alpha:0.6];
        [self.resetBtn setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
        [self.resetBtn removeTarget:self action:@selector(timerRecord) forControlEvents:UIControlEventTouchUpInside];
        
        [self.resetBtn addTarget:self action:@selector(timerReset) forControlEvents:UIControlEventTouchUpInside];
    }
}

- (void)timerReset {
    self.timerCount = 0;
    self.mainTimerLabel.text = @"00:00.00";
    self.currentCount = 0;
    self.maxCount = 0;
    self.minCount = 0;
    [self.dataArray removeAllObjects];
    
    self.resetBtn.enabled = NO;
    [self.resetBtn setTitle:@"计次" forState:UIControlStateDisabled];
    self.resetBtn.backgroundColor = [UIColor colorWithWhite:1 alpha:0.3];
    [self.resetBtn setTitleColor:[UIColor grayColor] forState:UIControlStateDisabled];
    [self.resetBtn removeTarget:self action:@selector(timerReset) forControlEvents:UIControlEventTouchUpInside];
    
    [self.resetBtn addTarget:self action:@selector(timerRecord) forControlEvents:UIControlEventTouchUpInside];
    
    [self.tableView reloadData];
}

- (void)timerRecord {
    if (self.currentCount > self.maxCount || self.maxCount == 0) self.maxCount = self.currentCount;
    if (self.currentCount < self.minCount || self.minCount == 0) self.minCount = self.currentCount;
    
    [self.dataArray insertObject:[NSNumber numberWithLong:self.currentCount] atIndex:0];
    
    self.currentCount = 0;
    
    [self.tableView reloadData];
    
}

//MARK: - tableview datasource
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.dataArray.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"CellId"];
    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleValue1 reuseIdentifier:@"CellId"];
    }
    
    long count = [self.dataArray[indexPath.row] longValue];
    
    cell.textLabel.text = [NSString stringWithFormat:@"计次 %ld", (long)(self.dataArray.count - indexPath.row)];
    cell.detailTextLabel.text = [NSString stringWithFormat:@"%.2ld:%.2ld.%.2ld", count/100/60, count/100%60, count%100];
    if (count >= self.maxCount) {
        cell.textLabel.textColor = [UIColor greenColor];
        cell.detailTextLabel.textColor = [UIColor greenColor];
        
    } else if (count <= self.minCount) {
        cell.textLabel.textColor = [UIColor redColor];
        cell.detailTextLabel.textColor = [UIColor redColor];
        
    } else {
        cell.textLabel.textColor = [UIColor whiteColor];
        cell.detailTextLabel.textColor = [UIColor whiteColor];
    }
    
    cell.backgroundColor = [UIColor blackColor];
    return cell;
}

//MARK: - tableview delegate
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 60;
}

//MARK: - view
- (void)configNavigationBar {
    self.navigationItem.title = @"秒表";
    
    self.navigationController.navigationBar.barTintColor = [UIColor blackColor];
    self.navigationController.navigationBar.titleTextAttributes = @{NSForegroundColorAttributeName:[UIColor whiteColor]};
}

- (void)configSubview {
    self.dataArray = [[NSMutableArray alloc] init];
    self.view.backgroundColor = [UIColor blackColor];
    
    self.mainTimerLabel = ({
        UILabel *label = [[UILabel alloc] init];
        label.font = [UIFont fontWithName:@"Arial Hebrew" size:68];
        label.textAlignment = NSTextAlignmentCenter;
        label.textColor = [UIColor whiteColor];
        label.text = @"00:00.00";
        label.adjustsFontSizeToFitWidth = YES;
        label;
    });
    [self.view addSubview:self.mainTimerLabel];
    self.mainTimerLabel.frame = CGRectMake(15, 100, [UIScreen mainScreen].bounds.size.width-30, [UIFont fontWithName:@"Arial Hebrew" size:68].lineHeight);
    
    self.resetBtn = ({
        UIButton *btn = [UIButton buttonWithType:UIButtonTypeCustom];
        btn.layer.cornerRadius = 40;
        btn.clipsToBounds = YES;
        [btn setTitle:@"计次" forState:UIControlStateDisabled];
        btn.backgroundColor = [UIColor colorWithWhite:1 alpha:0.3];
        [btn setTitleColor:[UIColor grayColor] forState:UIControlStateDisabled];
        btn.titleLabel.font = [UIFont systemFontOfSize:17];
        [btn addTarget:self action:@selector(timerRecord) forControlEvents:UIControlEventTouchUpInside];
        btn;
    });
    [self.view addSubview:self.resetBtn];
    self.resetBtn.frame = CGRectMake([UIScreen mainScreen].bounds.size.width/4-80/2, 250, 80, 80);
    self.resetBtn.enabled = NO;
    
    self.startBtn = ({
        UIButton *btn = [UIButton buttonWithType:UIButtonTypeCustom];
        btn.layer.cornerRadius = 40;
        btn.clipsToBounds = YES;
        [btn setTitle:@"开始" forState:UIControlStateNormal];
        btn.backgroundColor = [UIColor colorWithRed:0 green:1 blue:0 alpha:0.3];
        [btn setTitleColor:[UIColor greenColor] forState:UIControlStateNormal];
        btn.titleLabel.font = [UIFont systemFontOfSize:17];
        [btn addTarget:self action:@selector(timerStart) forControlEvents:UIControlEventTouchUpInside];
        btn;
    });
    [self.view addSubview:self.startBtn];
    self.startBtn.frame = CGRectMake([UIScreen mainScreen].bounds.size.width/4*3-80/2, 250, 80, 80);
    
    self.tableView = ({
        UITableView *tableView = [[UITableView alloc] initWithFrame:CGRectMake(0, 250+80+30, [UIScreen mainScreen].bounds.size.width, [UIScreen mainScreen].bounds.size.height-(250+80+30)) style:UITableViewStylePlain];
        tableView.separatorColor = [UIColor grayColor];
        tableView.separatorStyle = UITableViewCellSeparatorStyleSingleLine;
        tableView.delegate = self;
        tableView.dataSource = self;
        tableView.backgroundColor = [UIColor blackColor];
        tableView;
    });
    [self.view addSubview:self.tableView];
    
}

@end

File: Project1-Clock-Swift/Project1-Clock-Swift/ViewController.swift
//
//  ViewController.swift
//  Day1 Clock(swift)
//
//  Created by 杨警 on 2018/3/5.
//  Copyright © 2018年 杨警. All rights reserved.
//

import UIKit

class ViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
    
    var mainTimerLabel: UILabel!
    var startBtn: UIButton!
    var resetBtn: UIButton!
    var timer: DispatchSourceTimer?
    var timerCount: Int64 = 0
    var tableView: UITableView!
    var dataArray: Array<Int64> = []
    var currentCount: Int64 = 0
    var maxCount: Int64 = 0
    var minCount: Int64 = 0
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        self.configNavigationBar();
        self.configSubview();
    }
    
    //MARK: - timer
    @objc func timerStart() {
        startBtn.setTitle("停止", for: UIControlState.normal);
        startBtn.backgroundColor = UIColor.init(red: 1, green: 0, blue: 0, alpha: 0.3);
        startBtn.setTitleColor(UIColor.red, for: UIControlState.normal);
        startBtn.removeTarget(self, action: #selector(timerStart), for: UIControlEvents.touchUpInside);
        
        startBtn.addTarget(self, action: #selector(timerStop), for: UIControlEvents.touchUpInside);
        
        resetBtn.isEnabled = true;
        resetBtn.setTitle("计次", for: UIControlState.normal);
        resetBtn.backgroundColor = UIColor.init(white: 1, alpha: 0.6);
        resetBtn.setTitleColor(UIColor.white, for: UIControlState.normal);
        resetBtn.removeTarget(self, action: #selector(timerReset), for: UIControlEvents.touchUpInside);
        
        resetBtn.addTarget(self, action: #selector(timerRecord), for: UIControlEvents.touchUpInside);
        
        timer = DispatchSource.makeTimerSource(flags: [], queue: DispatchQueue.global());
        timer?.schedule(wallDeadline: DispatchWallTime.now(), repeating: 0.01);
        timer?.setEventHandler(handler: {
            self.timerCount += 1;
            self.currentCount += 1;
            
            DispatchQueue.main.async {
                self.mainTimerLabel.text = String.init(format: "%.2ld:%.2ld.%.2ld", self.timerCount/100/60, self.timerCount/100%60, self.timerCount%100);
            }
        });
        timer?.resume();
    }
    
    @objc func timerStop() {
        if (timer?.isCancelled) == false {
            timer?.cancel();
        }
        
        startBtn.setTitle("开始", for: UIControlState.normal);
        startBtn.backgroundColor = UIColor.init(red: 0, green: 1, blue: 0, alpha: 0.3);
        startBtn.setTitleColor(UIColor.green, for: UIControlState.normal);
        startBtn.removeTarget(self, action: #selector(timerStop), for: UIControlEvents.touchUpInside);
        
        startBtn.addTarget(self, action: #selector(timerStart), for: UIControlEvents.touchUpInside);
        
        if timerCount > 0 {
            resetBtn.setTitle("复位", for: UIControlState.normal);
            resetBtn.backgroundColor = UIColor.init(white: 1, alpha: 0.6);
            resetBtn.setTitleColor(UIColor.white, for: UIControlState.normal);
            resetBtn.removeTarget(self, action: #selector(timerRecord), for: UIControlEvents.touchUpInside);
            resetBtn.addTarget(self, action: #selector(timerReset), for: UIControlEvents.touchUpInside);
        }
    }
    
    @objc func timerReset() {
        timerCount = 0;
        mainTimerLabel.text = "00:00.00";
        currentCount = 0;
        maxCount = 0;
        minCount = 0;
        dataArray.removeAll();
        
        resetBtn.isEnabled = false;
        resetBtn.setTitle("计次", for: UIControlState.disabled);
        resetBtn.backgroundColor = UIColor.init(white: 1, alpha: 0.3);
        resetBtn.setTitleColor(UIColor.gray, for: UIControlState.disabled);
        resetBtn.removeTarget(self, action: #selector(timerReset), for: UIControlEvents.touchUpInside);
        resetBtn.addTarget(self, action: #selector(timerRecord), for: UIControlEvents.touchUpInside);
        
        tableView.reloadData();
    }
    
    @objc func timerRecord() {
        if currentCount > maxCount || maxCount == 0 {
            maxCount = currentCount;
        }
        
        if currentCount < minCount || minCount == 0 {
            minCount = currentCount;
        }
        
        dataArray.insert(currentCount, at: 0);
        currentCount = 0;
        
        tableView.reloadData();
    }
    
    //MARK: - tableView datasource
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return dataArray.count;
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        var cell: UITableViewCell? = tableView.dequeueReusableCell(withIdentifier: "CellId");
        if cell == nil {
            cell = UITableViewCell.init(style: UITableViewCellStyle.value1, reuseIdentifier: "CellId");
            cell?.backgroundColor = UIColor.black;
        }
        
        let count = dataArray[indexPath.row];
        
        cell!.textLabel?.text = "计次 \(dataArray.count-indexPath.row)";
        cell!.detailTextLabel?.text = String.init(format: "%.2ld:%.2ld.%.2ld", count/100/60, count/100%60, count%100);
        
        if count >= maxCount {
            cell?.textLabel?.textColor = UIColor.green;
            cell?.detailTextLabel?.textColor = UIColor.green;
            
        } else if count <= minCount {
            cell?.textLabel?.textColor = UIColor.red;
            cell?.detailTextLabel?.textColor = UIColor.red;
            
        } else {
            cell?.textLabel?.textColor = UIColor.white;
            cell?.detailTextLabel?.textColor = UIColor.white;
        }
        return cell!;
    }
    
    //MARK: - tableView delegate
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 60;
    }
    
    //MARK: - view
    func configNavigationBar() {
        navigationItem.title = "秒表";
        
        navigationController?.navigationBar.barTintColor = UIColor.black;
        navigationController?.navigationBar.titleTextAttributes = [NSAttributedStringKey.foregroundColor: UIColor.white];
    }
    
    func configSubview() {
        view.backgroundColor = UIColor.black;
        
        //mainTimerLabel
        mainTimerLabel = UILabel.init(frame: CGRect.init(x: 15, y: 100, width: UIScreen.main.bounds.size.width-30, height:UIFont.init(name:"Arial Hebrew", size:68)!.lineHeight));
        mainTimerLabel.font = UIFont.init(name:"Arial Hebrew", size:68);
        mainTimerLabel.textAlignment = NSTextAlignment.center;
        mainTimerLabel.textColor = UIColor.white;
        mainTimerLabel.text = "00:00.00";
        mainTimerLabel.adjustsFontSizeToFitWidth = true;
        view.addSubview(mainTimerLabel);
        
        //resetBtn
        resetBtn = UIButton.init(type: UIButtonType.custom);
        resetBtn.layer.cornerRadius = 40;
        resetBtn.clipsToBounds = true;
        resetBtn.backgroundColor = UIColor.init(white: 1, alpha: 0.3);
        resetBtn.setTitle("计次", for: UIControlState.normal);
        resetBtn.setTitleColor(UIColor.gray, for: UIControlState.normal);
        resetBtn.titleLabel?.font = UIFont.systemFont(ofSize: 17);
        resetBtn.addTarget(self, action:#selector(timerRecord) , for: UIControlEvents.touchUpInside);
        view.addSubview(resetBtn);
        resetBtn.frame = CGRect.init(x: UIScreen.main.bounds.size.width/4-80/2, y: 250, width: 80, height: 80);
        resetBtn.isEnabled = false;
        
        //startBtn
        startBtn = UIButton.init(type: UIButtonType.custom);
        startBtn.layer.cornerRadius = 40;
        startBtn.clipsToBounds = true;
        startBtn.setTitle("开始", for: UIControlState.normal);
        startBtn.backgroundColor = UIColor.init(red: 0, green: 1, blue: 0, alpha: 0.3);
        startBtn.setTitleColor(UIColor.green, for: UIControlState.normal);
        startBtn.titleLabel?.font = UIFont.systemFont(ofSize: 17);
        startBtn.addTarget(self, action: #selector(timerStart), for: UIControlEvents.touchUpInside);
        view.addSubview(startBtn);
        startBtn.frame = CGRect.init(x: UIScreen.main.bounds.size.width/4*3-80/2, y: 250, width: 80, height: 80);
        
        //tableView
        tableView = UITableView.init(frame: CGRect.init(x: 0, y: 250+80+30, width: UIScreen.main.bounds.size.width, height: UIScreen.main.bounds.size.height-(250+80+30)), style: UITableViewStyle.plain);
        tableView.separatorColor = UIColor.gray;
        tableView.separatorStyle = UITableViewCellSeparatorStyle.singleLine;
        tableView.delegate = self;
        tableView.dataSource = self;
        tableView.backgroundColor = UIColor.black;
        view.addSubview(tableView);
    }
    
}

